### ISSUE TITLE
Multiplicity in Domain type declarations

### ISSUE THREAD

**Issue by liangyiliang** [penrose/penrose#1726]
## Problem

Type declarations in Domain denote the types of objects that the Substance can reason about, but makes little restrictions on _how_ these types can be instantiated in Substance. One common use case is restricting the number of objects that can be associated with this type (subtyping included).

For example, if we would like to render a state of [the classic river-crossing example](https://alloytools.org/tutorials/online/frame-RC-1.html), we would want the Substance program diagram to consist of one chicken, one fox, one grain, and one farmer, no more and no less -- these are the multiplicity constraints. Currently, Domain does not allow us to encode these multiplicity constraints: 

```
type Farmer
type Fox
type Grain
type Chicken
```

Currently, the only way to enforce these multiplicity constraints are through Style, like
```
-- Style requires f1 and f2 to be disjoint
forall Farmer f1; Farmer f2 { 
  -- render an error message
}
```

This is unintuitive and just feels hacked together. Instead, Penrose should have built-in language support for these kinds of multiplicity constraints.

### Domain

Each type declaration in Domian should accept optional multiplicity declarations, like

```
one type TypeA
some type TypeB
lone type TypeC
-- etc ...
```

Here, `one` means there can and must be exactly one object of type `TypeA` in Substance; `some` means there must be at least one object of type `TypeB`, and `lone` (**l**ess-than-or-equal-to-**one**) means there must be zero or one object of type `TypeC` in Substance.

I personally like the keyword `one` but don't really like `some` and `lone` -- I use them here only because Alloy uses these keywords.

### Substance

The Substance compiler needs to check whether the Substance program respects these multiplicity constraints. In a perfect world this would be done by constructing a simple type-object map and checking that each type's list of objects has the permissable number of elements as dictated by the multiplicity constraints.

### Subtyping problems

One thing we need to consider is subtyping. As an example,
```
-- domain
one type SuperType
type SubType <: SuperType

-- substance
SubType a, b, c
```

This should be disallowed by the check since `SuperType` now essentially has three objects, violating its multiplicity constraints. We can of course do this check in Substance, but it _might_ make more sense to handle this in Domain, requiring `SubType`'s multiplicity to be compatible with `SuperType`'s multiplicity. We need to consider all the possible subtypes of a supertype, though:

```
-- domain
one type SuperType
one type SubType1 <: SuperType
one type SubType2 <: SuperType
```

This should be disallowed. Even though the multiplicities of both `SubType1` and `SubType2` both individually satisfy `SuperType`'s multiplicity, once appearing together, they violate it. One way is to define some sort of "arithmetic" to our very simple set of multiplicities:

```
Mult ::= one | some | lone | any

// `any` is the default multiplicity - no restrictions on multiplicity

one + one = some
one + some = some
one + lone = some
one + any = some

some + some = some
some + lone = some
some + any = some

lone + lone = any
lone + any = any

any + any = any
```

(I'm not sure if this is a good set of arithmetics ...)

Then, check whether or not the declared multiplicity of `SuperType` is consistent with the "sum" of the declared multiplicities of the `SubType`s.

**Comment by wodeni** [comment]
I'm generally in favor of giving Substance more semantics, and this idea seems cool. Whether this should be done through Style or Domain I'm a bit less sure. 

Out of curiosity, why does alloy have this in their type system? Was it mainly to constrain the solver so it produces (counter)examples respecting the multiplicity? Are there any languages that do this?

**Comment by liangyiliang** [comment]
> Out of curiosity, why does alloy have this in their type system? Was it mainly to constrain the solver so it produces (counter)examples respecting the multiplicity?

That's exactly why Alloy does this. In fact, 

```
one sig X {}
```

is equivalent to
```
sig X {}
fact {
  one X
}
```

The Alloy solver generates instances that respect all the `fact`s in the model.

**Comment by liangyiliang** [comment]
> Whether this should be done through Style or Domain I'm a bit less sure.

You mean Domain or Substance, right? This proposal has nothing to do with Style.

**Comment by wodeni** [comment]
> > Whether this should be done through Style or Domain I'm a bit less sure.
> 
> You mean Domain or Substance, right? This proposal has nothing to do with Style.

No, I meant Domain or Style because I was thinking about the alternative you mentioned: use pattern matching to catch errors in Style. It's a bit more expressive and flexible. I'm concerned about the case where different Style writers disagree on the semantics of the language. For example, if someone wants to adapt the classical river crossing diagram to illustrate stranger setups involving, say, multiple chickens, Domain will prevent compilation.

**Comment by liangyiliang** [comment]
> For example, if someone wants to adapt the classical river crossing diagram to illustrate stranger setups involving, say, multiple chickens, Domain will prevent compilation.

I see, interesting! But I'd argue that a problem domain that only allows one chicken is a different problem domain that would allow multiple chickens, so it should really be the domain writer's responsibility to make clear the scenarios that the domain would like to support.

In Penlloy, I imagine that the multiplicity constraint of a `sig` in the model can directly be translated into multiplicity constraint of a domain `type`. 

The idea is, the domain would respect the Alloy model, the style / substance must respect the domain and hence respect the Alloy model.

**Comment by liangyiliang** [comment]
## Abstract types

In Alloy, a signature can be declared as `abstract`. Like abstract classes in Java, which cannot be instantiated but can only be inherited, an abstract signature in Alloy cannot have any concrete objects, unless they are concretizations of its non-abstract sub-signatures.

As an example, in the river-crossing model, an object can either be a Chicken, a Fox, a Grain, or a Farmer. The four categories of objects are disjoint, and a object must belong to exactly one of the four categories. This would be declared as follows:

```alloy
abstract sig Object {}
one sig Chicken, Fox, Grain, Farmer extends Object {}
```

If we were to translate this into Penrose domain, it would be nice to be able to keep information such as `abstract` in addition to multiplicity constraints, something like
```
abstract type Object
one type Chicken
one type Fox
one type Grain
one type Farmer
```

The Substance checker would use the Domain program to check that there are no objects that belong to category `Object` but does not belong to any of its sub-categories.
